pkg/notifiers/sender.go

```diff
@@ -0,0 +1,302 @@
+package notifiers
+
+import (
+	"bytes"
+	"crypto/sha256"
+	"crypto/tls"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"github.com/prometheus/common/model"
+	"github.com/rancher/types/apis/management.cattle.io/v3"
+	"io/ioutil"
+	"mime/multipart"
+	"net"
+	"net/http"
+	"net/smtp"
+	"net/textproto"
+	"strconv"
+	"strings"
+	"time"
+)
+
+func SendMessage(notifier *v3.Notifier, recipient string, msg *Message) error {
+	if notifier.Spec.SlackConfig != nil {
+		if recipient == "" {
+			recipient = notifier.Spec.SlackConfig.DefaultRecipient
+		}
+		return TestSlack(notifier.Spec.SlackConfig.URL, recipient, msg.Content)
+	}
+
+	if notifier.Spec.SMTPConfig != nil {
+		s := notifier.Spec.SMTPConfig
+		if recipient == "" {
+			recipient = s.DefaultRecipient
+		}
+		return TestEmail(s.Host, s.Password, s.Username, int(s.Port), s.TLS, msg.Title, msg.Content, recipient, s.Sender)
+	}
+
+	if notifier.Spec.PagerdutyConfig != nil {
+		return TestPagerduty(notifier.Spec.PagerdutyConfig.ServiceKey, msg.Content)
+	}
+
+	if notifier.Spec.WebhookConfig != nil {
+		return TestWebhook(notifier.Spec.WebhookConfig.URL, msg.Content)
+	}
+
+	return nil
+}
+
+func TestPagerduty(key, msg string) error {
+	if msg == "" {
+		msg = "Pagerduty setting validated"
+	}
+
+	pd := &pagerDutyMessage{
+		ServiceKey:  key,
+		EventType:   "trigger",
+		IncidentKey: hashKey("key"),
+		Description: msg,
+	}
+
+	url := "https://events.pagerduty.com/generic/2010-04-15/create_event.json"
```

Comment:
would this url be permanent/static? or is it something that would be configurable

Reply:
It is static.
reference: https://v2.developer.pagerduty.com/docs/events-api#making-a-request